lib/Agrammon/OutputFormatter/PDF.pm6

```diff
@@ -133,30 +144,149 @@ sub input-output-as-pdf(
-        for @outputs -> %rec {
-            %rec<unit>  = latex-small-spaces(latex-escape(%rec<unit>));
-            %rec<label> = latex-chemify(latex-escape(%rec<label>));
-            %rec<value> = format-value(%rec<value>);
+        for @outputs.sort(+*.<order>) -> %rec {
+            my $print = %rec<print>; # can be undefined or empty
+            if $print and $print ne $last-print {
+                @output-formatted.push(%(
+                    :section(%lang-labels{$print}{$language} // 'NO-TITLE'),
+                    :$first));
+                $first = False if $first;
+                $last-print = $print;
+            }
+            @output-formatted.push(%(
+                :unit(latex-small-spaces(latex-escape(%rec<unit>))),
+                :label(latex-chemify(latex-escape(%rec<label>))),
+                :value(format-value(%rec<value>))));
+        }
+    }
+
+    my @input-formatted;
+    # TODO: fix sorting
+    my @inputs := %data<inputs>;
+# left on purpose
+#    dd @inputs[0];
+    $last-print = '';
+    my $last-instance = '';
+    my $last-module = '';
+    my $first-module = True;
+    my $first-instance = True;
+    for @inputs -> %rec {
+        my $new-module = False;
+        my $new-instance = False;
+
+        my $module = %rec<gui>;
+        if $module ne $last-module {
+            @input-formatted.push( %(
+                :module(latex-escape($module)),
+                :$first-module));
+            $last-module = $module;
+            $new-module = True;
+            $first-module = False;
+            $first-instance = True;
         }
+
+        # instance can be empty for none-multi modules
+        my $instance = %rec<instance>;
+        if $instance and $instance ne $last-instance {
+            @input-formatted.push( %(
+                :instance(latex-escape($instance)),
+                :$first-instance));
+            $last-instance = $instance;
+            $new-instance = True;
+            $first-instance = False;
+        }
+
+        my $first-line = $new-module && ! $new-instance;
+
+        @input-formatted.push(%(
+            :unit(latex-small-spaces(latex-escape(%rec<unit>))),
+            :label(latex-chemify(latex-escape(%rec<input>))),
+            :value(format-value(%rec<value-translated>)),
+            :$first-line));
     }
-    %data<dataset>  = $dataset-name;
-    %data<username> = $user.username;
-    %data<model>    = $cfg.gui-variant;
 
-    my $pdf = create-pdf(
+    # setup template data
+    %data<outputs>   = @output-formatted;
+    %data<inputs>    = @input-formatted;
+    %data<titles>    = %titles;
+    %data<dataset>   = $dataset-name;
+    %data<username>  = $user.username;
+    %data<model>     = $cfg.gui-variant;
+    %data<version>   = 'Agrammon 6.0';
+    %data<timestamp> = ~DateTime.now( formatter => sub ($o) {
+        sprintf '%02d.%02d.%04d %02d:%02d:%02d',
+            $o.day, $o.month, $o.year,
+            $o.hour, $o.minute, $o.second,
+    });
```

Comment:
Really isn't important, but it can be less repetitive to do:

```
@jnthn	Reply…
    %data<timestamp> = ~DateTime.now( formatter => sub ($_) {
        sprintf '%02d.%02d.%04d %02d:%02d:%02d',
            .day, .month, .year,
            .hour, .minute, .second,
    });
```

Reply:
Fixed.